### PR TITLE
Fix/better coverage

### PR DESCRIPTION
**Pull request by solarw**
## Proposed changes

better code coverage with tests

## Fixes

coverage gaps

## Types of changes

- [x] Bugfix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Checklist

- [x] I have read the [CONTRIBUTING](../CONTRIBUTING.md) doc
- [x] I am making a pull request against the `develop` branch (left side). Also you should start your branch off our `develop`.
- [x] Lint and unit tests pass locally with my changes and CI passes too
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have checked that code coverage does not decrease.
- [x] I have added necessary documentation (if appropriate)
- [x] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

**Comment by Karrenbelt** [comment]
@solarw `git cherry-pick 91284faf9f2c94b2c4a40093585eeb999fc48267` - see 91284faf9f2c94b2c4a40093585eeb999fc48267

File: aea/package_manager/base.py [hunk 1]
```diff
@@ -181,6 +180,30 @@ def check_dependencies(
...
 
+    def is_dependencies_hashes_match(
+        self, package_id: PackageId, configuration_obj: PackageConfiguration
+    ) -> bool:
+        """Check dependecies hashes match and print errors"""
+        dependencies_with_mismatched_hashes = self.check_dependencies(
+            configuration=configuration_obj,
+        )
+        if len(dependencies_with_mismatched_hashes) > 0:
+            for dep, failure in dependencies_with_mismatched_hashes:
```
Comment (Karrenbelt) [inline aea/package_manager/base.py:191]: the if check is not necessary here, can iterate an empty list

then `return bool(dependencies_with_mismatched_hashes)` at the end

File: aea/package_manager/v1.py [hunk 2]
```diff
@@ -314,6 +296,11 @@ def json(
 
         return data
 
+    @staticmethod
+    def _load_packages(packages_file: Path) -> Dict:
+        """Load packages json file."""
+        return cast(Dict, json.loads(packages_file.read_text()))
```
Comment (Karrenbelt) [inline aea/package_manager/v1.py:302]: `Dict[str, Dict[str, str]]`

File: packages/fetchai/connections/http_server/tests/test_http_server.py [hunk 3]
```diff
@@ -268,7 +269,7 @@ async def test_bad_performative_get_timeout_error(self):
     @pytest.mark.asyncio
     async def test_late_message_get_timeout_error(self):
         """Test send get request w/ 200 response."""
-        self.http_connection.channel.timeout_window = 1
+        self.http_connection.channel.timeout_window = 0.2
```
Comment (Karrenbelt) [inline packages/fetchai/connections/http_server/tests/test_http_server.py:272]: can you explain? particularly why `a < b` now instead of `a > b` as it was

File: packages/valory/connections/ledger/tests/test_contract_dispatcher.py [hunk 4]
```diff
@@ -0,0 +1,280 @@
...
+    """Tests a default method call through ContractApiRequestDispatcher."""
+
+    dummy_address = "0x0000000000000000000000000000000000000000"
+    contract_dispatcher = ContractApiRequestDispatcher(
+        connection_id=Mock(), connection_state=AsyncState()
+    )
+    contract = Mock()
+
+    def callable(ledger_api=None, contract_address=None, _addr=None):  # type: ignore
+        return 0
```
Comment (Karrenbelt) [inline packages/valory/connections/ledger/tests/test_contract_dispatcher.py:94]: overwrite of built-in function `callable`, please rename

File: tests/test_helpers/test_env_vars.py [hunk 5]
```diff
@@ -195,3 +197,20 @@ def test_match_export_parse_consistency(export_data, template) -> None:
...
+    )
+    assert result == [{"arg": 12}, {"arg": 12, "public_id": "fetchai/test:0.1.0"}]
+
+
+def test_is_strict_list():
+    """Test is_strict method."""
+    assert is_strict_list([1, 2, 3])
+    assert is_strict_list([1, 2, 3, [1, 2, 3]])
+    assert not is_strict_list([1, 2, {}])
+    assert not is_strict_list([1, 2, [[{}]]])
```
Comment (Karrenbelt) [inline tests/test_helpers/test_env_vars.py:216]: the original function is not clear and should likely not exist. 

is_strict_list([tuple(dict(hello="world"), )])
>>> True

File: packages/valory/connections/ledger/tests/test_contract_dispatcher.py (hunk 4 above)
Comment (solarw) [reply packages/valory/connections/ledger/tests/test_contract_dispatcher.py:94]: done

File: packages/fetchai/connections/http_server/tests/test_http_server.py (hunk 3 above)
Comment (solarw) [reply packages/fetchai/connections/http_server/tests/test_http_server.py:272]: just decreased timeout and next sleep. make it a bit faster. looks ridiculous, but many sleeps cosnumea lot of time.
it's better to get rid of all sleeps in tests.

File: aea/package_manager/v1.py (hunk 2 above)
Comment (solarw) [reply aea/package_manager/v1.py:302]: done. thanks good point

File: aea/package_manager/base.py (hunk 1 above)
Comment (solarw) [reply aea/package_manager/base.py:191]: brilliant. done

File: tests/test_helpers/test_env_vars.py (hunk 5 above)
Comment (solarw) [reply tests/test_helpers/test_env_vars.py:216]: good point.
for me, it looks like a check for nested dicts

added support of tuple

```assert is_strict_list([tuple(dict(hello="world"), )])``` - is true, tuple(dict) => tuple of dict keys

Comment (Karrenbelt) [reply tests/test_helpers/test_env_vars.py:216]:
```
def is_container(obj: Any) -> bool:
    return isinstance(obj, (MutableSequence, MutableSet, MutableMapping))

def is_strict_list(data: List):
    for c in filter(is_container, data):
        if isinstance(c, list):
            return is_strict_list(c)
        return False
    return True
```

Comment (solarw) [reply tests/test_helpers/test_env_vars.py:216]: good point and nice to have a ticket with proposals.

File: aea/package_manager/v1.py [hunk 6]
```diff
@@ -219,6 +218,11 @@ def update_package_hashes(self) -> "PackageManagerV1":
 
         return self
 
+    @staticmethod
+    def _calculate_hash(package_path: str) -> str:
+        """Calculate hash for path."""
+        return IPFSHashOnly.get(str(package_path))
```
Comment (Karrenbelt) [inline aea/package_manager/v1.py:224]: not worth a method, please don't because this is all interface we will have to preserve into the future for reasons of compatibility :cry:

File: aea/package_manager/v1.py [hunk 7]
```diff
@@ -314,6 +296,11 @@ def json(
 
         return data
 
+    @staticmethod
+    def _load_packages(packages_file: Path) -> Dict[str, Dict[str, str]]:
+        """Load packages json file."""
+        return cast(Dict, json.loads(packages_file.read_text()))
```
Comment (Karrenbelt) [inline aea/package_manager/v1.py:302]: same thing

File: aea/runtime.py [hunk 8]
```diff
@@ -346,7 +346,7 @@ async def run_runtime(self) -> None:
         """Run runtime which means start agent loop, multiplexer and storage."""
         self._state.set(RuntimeStates.starting)
         await asyncio.gather(
-            self._start_multiplexer(), self._start_agent_loop(), self._start_storage()
+            self._start_agent_loop(), self._start_multiplexer(), self._start_storage()
```
Comment (Karrenbelt) [inline aea/runtime.py:349]: why the change of order?

File: tests/test_package_manager/test_v1.py [hunk 9]
```diff
@@ -105,6 +126,59 @@ def test_sync(
...
+        with mock.patch.object(pm, "add_package") as add_patch, mock.patch.object(
+            pm, "_update_package"
+        ) as update_patch, mock.patch(
+            "pathlib.Path.exists", return_value=True
+        ), mock.patch(
+            "aea.helpers.ipfs.base.IPFSHashOnly.get",
+            return_value=DUMMY_PACKAGE_HASH + "X",
+        ):
+            pm.sync(dev=True, third_party=False, update_packages=True)
+            update_patch.assert_called_once()
```
Comment (Karrenbelt) [inline tests/test_package_manager/test_v1.py:179]: there is a tiny context manager stacking tool, but it's not quite optimal. We can wrap it to make it nice so we can stack context managers in a way code is actually readable. 
https://github.com/valory-xyz/open-aea/blob/6e6a444ce2fee62932031f1bffdae58f24f52aeb/aea/test_tools/utils.py#L67-L72

I will add a tiny bit to this soon:tm: 
1. `yield stack`
2. we can then access stacked context managers by index (e.g. `stack.args[0]`). I want to wrap it so that we can initialize it with kwargs. E.g. 
```
with as_context(
    a=..., 
    b=...,
) as m: 
    ...
    m.a.assert_called()
    m.b.assert_not_called()
```

@0xArdi what do you think?

File: tests/test_package_manager/test_v1.py [hunk 10]
```diff
@@ -407,3 +565,34 @@ def test_package_manager_add_package_can_be_updated(fetch_mock: mock.Mock):
 
             package_manager.add_package(TEST_SKILL_ID, allow_update=True)
             remove_mock.assert_called_once_with(TEST_SKILL_ID)
+
+
+class TestVariousMEthods(BaseAEATestCase):
```
Comment (Karrenbelt) [inline tests/test_package_manager/test_v1.py:570]: unimportant: `TestVariousMEthods` -> `TestVariousMethods`

File: tests/test_runtime.py [hunk 11]
```diff
@@ -120,6 +120,21 @@ def test_double_stop(self):
...
+        task = asyncio.ensure_future(self.runtime.run_runtime())
+        await wait_for_condition_async(lambda: self.runtime.is_running, timeout=5)
+        assert self.runtime.agent_loop.is_running
+        task.cancel()
+        await asyncio.sleep(0.1)
+
+        with pytest.raises(asyncio.CancelledError):
+            await task
+
+        assert not self.runtime.agent_loop.is_running
```
Comment (Karrenbelt) [inline tests/test_runtime.py:136]: just curious since you said we should remove all sleep. How would you remove it here? I'm thinking we could combine `wait_for_condition` pattern with a signal handler. Is there a better or easier way?

File: tests/test_runtime.py [hunk 12]
```diff
@@ -173,3 +188,8 @@ def test_error_state(self):
                 self.runtime.start_and_wait_completed(sync=True)
 
         assert self.runtime.state == RuntimeStates.error, self.runtime.state
+
+    @pytest.mark.asyncio
+    async def test_runtime_cancelled(self):
+        """Test runtime cancelled."""
+        pytest.skip("only for async runtime")
```
Comment (Karrenbelt) [inline tests/test_runtime.py:195]: this is... just very odd. Can this be removed or does it serve some esoteric purpose? If in fact the latter please explicate.

File: tests/test_package_manager/test_v1.py (hunk 10 above)
Comment (solarw) [reply tests/test_package_manager/test_v1.py:570]: done

File: aea/package_manager/v1.py (hunk 6 above)
Comment (solarw) [reply aea/package_manager/v1.py:224]: @Karrenbelt 
actually i like such kind of utlility methods that interacts with some external things

in case you want to change hash calc tool/algo, just override the method without looking over the code.
so this makes sense for me

Comment (solarw) [reply aea/package_manager/v1.py:224]: and easy to mock)

File: aea/package_manager/v1.py (hunk 7 above)
Comment (solarw) [reply aea/package_manager/v1.py:302]: @Karrenbelt same here:
if we want to move from json to yaml, it stays incapsulated here and easy to mock

File: aea/runtime.py (hunk 8 above)
Comment (solarw) [reply aea/runtime.py:349]: actually was done during some exploration, nothing changed actually.

was looking for unreachable code and wondered is order important. can be reverted, or not.

File: tests/test_package_manager/test_v1.py (hunk 9 above)
Comment (solarw) [reply tests/test_package_manager/test_v1.py:179]: interesting, but suppose can just reformat this block of `with ... as ...` better and make it more readable

File: tests/test_runtime.py (hunk 11 above)
Comment (solarw) [reply tests/test_runtime.py:136]: @Karrenbelt
my bad. it ws experimental thing during unreachable code  investigation. forgot to remove

File: tests/test_runtime.py (hunk 12 above)
Comment (solarw) [reply tests/test_runtime.py:195]: updated properly

File: aea/package_manager/v1.py (hunk 6 above)
Comment (Adamantios) [reply aea/package_manager/v1.py:224]: Either `package_path`'s type hint is incorrect or `str(package_path)` is redundant.

Comment (Karrenbelt) [reply aea/package_manager/v1.py:224]: > @Karrenbelt actually i like such kind of utlility methods that interacts with some external things
> 
> in case you want to change hash calc tool/algo, just override the method without looking over the code. so this makes sense for me

ok I don't like to expose any more interface than is absolutely necessary. 1. Interface that we now need to maintain, and 2. it adds another alias for an existing method (which is the lesser of these two evils).

Let me recommend to do it this way then:
```
class PackageManagerV1(BasePackageManager):
    """Package manager V1"""
    
    get_ipfs_hash = IPFSHashOnly.get
    
```

File: aea/runtime.py (hunk 8 above)
Comment (Karrenbelt) [reply aea/runtime.py:349]: okay, it's fine then

File: tests/test_runtime.py (hunk 12 above)
Comment (Karrenbelt) [reply tests/test_runtime.py]: I don't understand this test though. What does it do? What is its purpose?

File: tests/test_runtime.py (hunk 11 above)
Comment (solarw) [reply tests/test_runtime.py]: done

File: aea/package_manager/v1.py (hunk 6 above)
Comment (solarw) [reply aea/package_manager/v1.py]: fixed

Comment (solarw) [reply aea/package_manager/v1.py]: @Karrenbelt 
tried your variant
```
aea/package_manager/v1.py:237: error: Invalid self argument "PackageManagerV1" to attribute function "_get_ipfs_hash" with type "Callable[[str, bool, bool], str]"
aea/package_manager/v1.py:237: error: Argument 1 has incompatible type "str"; expected "bool"
```

dont want to investigate why it does not work for mypy, tests are ok.

so left it as is.

actually good point to discuss.

File: aea/runtime.py (hunk 8 above)
Comment (DavidMinarsch) [reply aea/runtime.py:349]: @solarw since we don't 100% know that this won't affect some edge case in prod I'd say we retain the original order

File: aea/runtime.py [hunk 13]
```diff
@@ -389,12 +389,8 @@ async def _start_agent_loop(self) -> None:
...
 
         self._state.set(RuntimeStates.running)
-        try:
-            await self.agent_loop.wait_completed()
-        except asyncio.CancelledError:
-            self.agent_loop.stop()
-            await self.agent_loop.wait_completed()
-            raise
+        # loop will be stopped in wait completed
+        await self.agent_loop.wait_completed()
```
Comment (DavidMinarsch) [inline aea/runtime.py:393]: Can you point me to the code?

Comment (solarw) [reply aea/runtime.py:393]: **https://github.com/valory-xyz/open-aea/blob/main/aea/agent_loop.py#L124**

so i captures cancelled event and stops agent loop.
checked it.

File: aea/runtime.py (hunk 8 above)
Comment (solarw) [reply aea/runtime.py]: @DavidMinarsch restored

**Review by DavidMinarsch — APPROVED** [review]
